README.md
# Abandoned

Do not use. Go [here](https://github.com/pimzand/jack) instead.

# Jack - Rip and encode CDs with one command

Jack has been developed with one main goal: making OGGs (or MP3s)
without having to worry. There is nearly no way that an incomplete rip
goes unnoticed, e.g. jack compares WAV and OGG file sizes when
continuing from a previous run. Jack also checks your HD space before
doing anything (even keeps some MB free).

Jack is different from other such tools in a number of ways:
- it supports different rippers and encoders
- it is very configurable
- it doesn't need X
- it can "rip" virtual CD images like the ones created by cdrdao
- when using cdparanoia, cdparanoia's status information is displayed and archived for all tracks, so you can see if something went wrong
- it uses sophisticated disk space management, i.e. it schedules its ripping/encoding processes depending on available space.
- gnudb query, file renaming and id3/ogg-tagging
- it can resume work after it has been interrupted. If all tracks have been ripped, it doesn't even need the CD anymore, even if you want to do a gnudb query.
- it can do a gnudb query based on OGGs alone, like if you don't remember from which CD those OGGs came from.
- gnudb submissions

## News

Jack is back! After a small 15 year hiatus, development has resumed.

Many thanks to the [friendly Debian maintainers](https://github.com/zzarne/jack/blob/master/debian/copyright) who kept Jack alive all these years!

## Goals

Current development goals include:
- Test and complete support for Python 3
- Get Jack back into distributions which no longer support Python 2
- Get Jack into PyPI
- Get the character encodings right. It's been a mess in old Python and Jack
- Rework the helpers and plugins into separate Python modules


debian/NEWS
jack (3.1.1-14) unstable; urgency=low

  As of this version, base_dir is no longer set in /etc/jackrc.  The
  reason for this is that there is no standard location for sound or
  media files and users may be surprised to find their files in ~/jack
  (the previous default of base_dir).  jack will now by default put
  the files in the current working directory.  You're welcome to set
  base_dir in ~/.jack3rc (or /etc/jackrc) if you have a more appropriate
  location.

 -- Martin Michlmayr <tbm@cyrius.com>  Wed, 11 Jan 2006 13:51:32 +0000

jack (3.0.0-1) experimental; urgency=low

  The configuration has changed fundamentally with jack-3.0.0.
  There is now a system-wide configuration file /etc/jackrc and
  the user configuration files have been renamed to ~/.jack3rc.
  Further, jack now has a simple scheme to set options permanently,
  by adding --save to the command-line. This is the preferred way
  to set options, as the syntax of the jackrc file has changed.
  See the commented out options in /etc/jackrc for examples of
  useful options and how to set them.

 -- Michael Banck <mbanck@debian.org>  Sat, 28 Feb 2004 21:30:43 +0000

jack (2.99.9-6) unstable; urgency=low

  jack now features a '--cddb-id' option, which lets you specify a
  freedb CDDB id, even though jack calculates a different one. You
  need to specify an ID3 genre via '-G/--id3-genre' for the freedb
  lookup, too. Note that this option has not been applied or even
  accepted upstream yet and is suspect to change in a later version.

 -- Michael Banck <mbanck@debian.org>  Sun, 31 Aug 2003 00:52:11 +0200

jack (2.99.9-2) unstable; urgency=low

  Alban Browaeys has pointed out that invalid OGG Vorbis tags are produced
  under certain circumstances.  While the OGG Vorbis specification requires
  that the field contents is encoded in UTF-8, jack up to 2.99.9-1 wrote
  an ASCII string to the field.  This has been fixed in 2.99.9-2.
  Fortunately, you do not have to re-encode your OGG files.  Rather, you
  can use jack's -R (rename) feature to simply update the OGG Vorbis tags.
  A script is supplied which takes care of this automatically.  Simply change
  to the directory you keep your OGG files in, and run the script.  For
  example, if your OGG files are in /media/ogg, you have to do:

    $ cd /media/ogg
    $ /usr/share/doc/jack/fix-vorbis-tags.pl

 -- Martin Michlmayr <tbm@cyrius.com>  Sun, 22 Dec 2002 21:53:22 +0100

jack (2.99.9-1) unstable; urgency=low

  Jack now features an undocumented option to turn off
  cdparanoia's --abort-on-skip option. The option's name is
  --I-swear-I'll-never-give-these-files-to-anyone,-including-hot-babes-TM
  Only use this option (if at all) when you already know that a CD will
  not get ripped OK, as this means that unsuccesful rips will be
  unnoticed. You have been warned and will continue on your own risk.

 -- Michael Banck <mbanck@debian.org>  Tue,  3 Sep 2002 02:26:19 +0200

jack (2.99.8-2) unstable; urgency=low

  It is now preferred to specify the vbr_quality rather than a
  bitrate. You can do this by adding vbr_quality = $QUALITY to
  your ~/.jackrc, $QUALITY being an integer from 0 (worst
  quality, smallest files) to 10 (best quality, biggest files)

  Average bitrates for oggenc will be around 70-100 for 0 up to
  300-500 kbit/s for 10. vbr_quality = 5 (the default if you don't
  specify the switch in your ~/.jackrc) will be around 150-170
  kbit/s. Please let me know if you get completely different bit-
  rates so I can tune the default. Note that lame and gogo do
  support this switch, too, but at least lame will have a much
  slimmer bandwidth for the bitrate (from 100-200 kbit/s) than
  oggenc.

  If you want to keep using the bitrate (I've been told that the
  quality would be inferior at the same filesize), you can switch
  the comments in /usr/bin/jack, line 425/426, or add a new helper
  section like helpers['oggenc-local'] = {'type':"encoder"[...]"""}
  to your ~/.jackrc(see end of auto-generated ~/.jackrc for an example)

 -- Michael Banck <mbanck@debian.org>  Fri, 24 May 2002 20:40:07 +0200

jack (2.99.7-4) unstable; urgency=low

  base_dir is now set to 'environ['HOME'] + "/jack"' (i.e. ~/jack)
  when a new .jackrc is created. This is in preparation for a jack
  menu item, so that jack does not dump the OGGs in the directory
  you started X from. If you upgrade, nothing should change ATM.

 -- Michael Banck <mbanck@gmx.net>  Sat, 23 Feb 2002 13:16:34 +0100

jack (2.99.7-1) unstable; urgency=low

  Jack-2.99.7 had a new .jackrc layout, you'll have to move your
  old .jackrc elsewhere and tune the newly generated one.

  If jack fails reading the TOC(table of contents) for your
  CD-drive, you can try adding
    toc_prog = "cdda2wav"
  or
    toc_prog = "cdparanoia"
  or simply
    toc_prog = ripper
  manually in your .jackrc.

  But beware, using this is deprecated and should be a last resort.
  In particular, you will lose the ability to rip CDs with data tracks,
  as jack no longer will be able to compare the TOCs!

  Please add to bug-nr. #118731 if you experience problems with
  your CD-drive.

 -- Michael Banck <mbanck@gmx.net>  Mon, 08 Oct 2001 18:02:20 +0000



jack/mp3.py
# jack.mp3 - mp3 layer stuff for
# jack - extract audio from a CD and MP3ify it using 3rd party software
# Copyright (C) 1999-2001  Arne Zellentin <zarne@users.sf.net>
# Transformed from mp3info (c) Thorvald Natvig

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

from functools import reduce
from io import StringIO
from stat import ST_SIZE
from os import stat
DEBUG = 0


global warnings
global F, f, xing, id3v2, start_offset

# constants
mode_names = ["stereo", "j-stereo", "dual-ch", "single-ch", "multi-ch"]
layer_names = ["I", "II", "III"]
version_names = ["MPEG-1", "MPEG-2 LSF", "MPEG-2.5"]
version_nums = ["1", "2", "2.5"]
bitrates = \
    [[[0, 32, 64, 96, 128, 160, 192, 224, 256, 288, 320, 352, 384, 416, 448],
        [0, 32, 48, 56, 64, 80, 96, 112, 128, 160, 192, 224, 256, 320, 384],
        [0, 32, 40, 48, 56, 64, 80, 96, 112, 128, 160, 192, 224, 256, 320]],
     [[0, 32, 48, 56, 64, 80, 96, 112, 128, 144, 160, 176, 192, 224, 256],
        [0, 8, 16, 24, 32, 40, 48, 56, 64, 80, 96, 112, 128, 144, 160],
        [0, 8, 16, 24, 32, 40, 48, 56, 64, 80, 96, 112, 128, 144, 160]],
     [[0, 32, 48, 56, 64, 80, 96, 112, 128, 144, 160, 176, 192, 224, 256],
        [0, 8, 16, 24, 32, 40, 48, 56, 64, 80, 96, 112, 128, 144, 160],
        [0, 8, 16, 24, 32, 40, 48, 56, 64, 80, 96, 112, 128, 144, 160]]]
s_freq = \
    [[44100, 48000, 32000, 0],
     [22050, 24000, 16000, 0],
     [11025, 8000, 8000, 0]]

bpf_fact = [0, 12000, 144000, 144000]

# Xing stuff
FRAMES_FLAG = 0x0001
BYTES_FLAG = 0x0002
TOC_FLAG = 0x0004
VBR_SCALE_FLAG = 0x0008


def get_l4(s):
    return reduce(lambda a, b: ((a << 8) + b), list(map(int, list(map(ord, s)))))


def decode_xing():
    global F, f, xing
    # defaults:
    frames = size = scale = toc = None
    where = f.tell()
    try:
        if 1:
            if DEBUG:
                print("Xing at", where)
            # 32-bit fields; "Xing", flags, frames, size, 100 toc
            flags = get_l4(f.read(4))
            if flags & FRAMES_FLAG:
                frames = get_l4(f.read(4))
            if flags & BYTES_FLAG:
                size = get_l4(f.read(4))
            if flags & TOC_FLAG:
                toc = [ord(x) for x in f.read(100)]
                for j in range(len(toc)):
                    toc[j] = int(toc[j] / 256.0 * size)
            if flags & VBR_SCALE_FLAG:
                scale = get_l4(f.read(4))

            if DEBUG > 2:
                print("Xing: frames=%i" % frames)
                print("Xing: bytes=%i" % size)
                print("Xing: toc=", toc)
                print("Xing: scale=%i" % scale)

            xing = {}
            if frames:
                xing['x_frames'] = frames
            if size:
                xing['x_bytes'] = size
            if toc:
                xing['x_toc'] = toc
            if scale:
                xing['x_scale'] = scale
            xing['x_header_at'] = where - 4

        else:
            return None

    except:
        if DEBUG:
            import traceback
            traceback.print_exc()
        f.seek(where + 1)
        return None


def find_xing(framelen):
    "read (and skip) the Xing header if present"
    where = f.tell()
    x = f.read(framelen)
    pos = x.find("Xing")
    if pos >= 0:
        f.seek(where + pos + 4)
        decode_xing()
        f.seek(where + framelen)
    else:
        f.seek(where)


def reread_f(size):
    global F, f
    pos = f.tell()
    f.seek(F.offset)
    if DEBUG:
        print("reread_f: now at", f.tell())
    F_offset = F.offset
    F_file = F.file
    F = StringIO(f.read(F.len + size + 10))
    F.file = F_file
    F.offset = F_offset
    f.seek(pos)


def handle_id3v2(version, flags, data):
    global id3v2
    if DEBUG:
        print("ID3:", repr(version), repr(flags), len(data))
    id3v2 = {}
    id3v2['version'] = tuple(version)


def read_id3v2():
    # $49 44 33 yy yy xx zz zz zz zz ; yy < $FF; zz < $80
    # I  D  3  yy yy xx zz zz zz zz
    # 0  1  2  3  4  5  6  7  8  9
    global F, f
    where = f.tell()
    header = f.read(10)
    if header[:3] != "ID3":
        if DEBUG:
            print("no header?", where, repr(header))
        f.seek(where)
        return None

    header = list(header)
    for i in range(3, 10):
        header[i] = ord(header[i])

    if not (header[3] < 0xff) or not (header[4] < 0xff):
        f.seek(where)
        return None
    if not (header[6] < 0x80) or not (header[7] < 0x80):
        f.seek(where)
        return None
    if not (header[8] < 0x80) or not (header[9] < 0x80):
        f.seek(where)
        return None

    size = header[6:]
    size = list(size)
    value = 0
    for i in size:
        value = value << 7
        value = value | i
    size = value

    reread_f(size)

    flags = header[5]
    version = header[3:5]
    return handle_id3v2(version, flags, f.read(size))


def decode_header(buffer):
    "get mp3 info. Transformed from mp3info (c) Thorvald Natvig"

    try:
        switch = (ord(buffer[1]) >> 3 & 0x3)
        if switch == 3:
            version = 0
        elif switch == 2:
            version = 1
        elif switch == 0:
            version = 2
        else:
            return None

        bitrate_index = (ord(buffer[2]) >> 4) & 0x0F
        sampling_frequency = (ord(buffer[2]) >> 2) & 0x3
        mode = (ord(buffer[3]) >> 6) & 0x3

        x = {}
        x['error_protection'] = not (ord(buffer[1]) & 0x1)
        x['padding'] = (ord(buffer[2]) >> 1) & 0x01
        x['extension'] = ord(buffer[2]) & 0x01
        x['mode_ext'] = (ord(buffer[3]) >> 4) & 0x03
        x['copyright'] = (ord(buffer[3]) >> 3) & 0x01
        x['original'] = (ord(buffer[3]) >> 2) & 0x1
        x['emphasis'] = (ord(buffer[3])) & 0x3
        x['mode_name'] = mode_names[mode]
        x['lay'] = 4 - ((ord(buffer[1]) >> 1) & 0x3)
        x['layer_name'] = layer_names[x['lay'] - 1]
        x['version_name'] = version_names[version]
        x['version_num'] = version_nums[version]
        x['bitrate'] = bitrates[version][x['lay'] - 1][bitrate_index]
        if not x['bitrate']:
            raise IllegalError_may_be_free_format
        x['sfreq'] = s_freq[version][sampling_frequency]
        if not x['sfreq']:
            raise IllegalError_illegal_sfreq

    except:
        if DEBUG:
            import traceback
            traceback.print_exc()
            print("Illegal header, so far:", x)
        x = None

    return x


def extend_frameinfo(x):
    if x['version_num'] == "1":
        x['samples_per_frame'] = 1152
    elif x['version_num'] == "2":
        x['samples_per_frame'] = 576
    else:
        print("What is the spf for %s?" % x['version_name'], "(%s)" % x['version_num'])

    # calculate BPF
    x['bpf'] = (x['bitrate'] * bpf_fact[x['lay']]) // x['sfreq']
    x['framesize'] = (x['bitrate'] * bpf_fact[x['lay']]) // x['sfreq'] + x['padding']
    # stereo = (mode == MPG_MD_MONO) ? 1 : 2;
    return x


def syncronize(what=1):
    global F, f
    index = f.tell()
    x = {}
    # search intelligently (i.e. fast) for any header
    fields = ["\377", "I", "X", "R"][:what]
    if not warnings:
        fields.remove("R")
    # while index < len(f.buf) - 4:
    while index < F.len - 4 + start_offset:
        # print index
        if F.buf[index - start_offset] in fields:
            if F.buf[index - start_offset + 1] >= "\340":
                # a valid header must start with 11 set bits
                f.seek(index)
                x = decode_header(f.read(4))
                if x:
                    where = f.tell() - 4
                    x['header_at'] = where
                    x = extend_frameinfo(x)
                    next_header = where + x['framesize']
                    if DEBUG:
                        print("MPEG header found at %i, expect next header at %i." % (x['header_at'], next_header))
                    f.seek(next_header)
                    buf = f.read(4)
                    y = None
                    if buf[:2] >= "\377\340":
                        y = decode_header(buf)
                    if y:
                        # two headers, that's certain enough.
                        f.seek(x['header_at'])
                        break
                    else:
                        if DEBUG:
                            print("but no valid header follows at %i." % next_header)
                        f.seek(where)
                        index = where
                        x = None
            elif F.buf[index - start_offset:index - start_offset + 3] == "ID3":
                if not id3v2:
                    f.seek(index)
                    read_id3v2()
                    if id3v2:
                        if DEBUG:
                            print("ID3v2 found, now at %i" % f.tell(), id3v2)
                        index = f.tell()
                        fields.remove("I")
                        continue
                    else:
                        f.seek(index)
            elif F.buf[index - start_offset:index - start_offset + 4] == "Xing":
                f.seek(index + 4)
                decode_xing()
                if xing:
                    if DEBUG:
                        print("Xing header found without frame sync.")
                    index = f.tell()
                    fields.remove("X")
                    continue
                else:
                    f.seek(index + 4)

            elif F.buf[index - start_offset:index - start_offset + 4] == "RIFF":
                if warnings:
                    print("skipping RIFF header at %i..." % F.tell())
                fields.remove("R")
        index = index + 1
    return x


def mp3format(file, warn=1, max_skip=100000, offset=0):
    global F, f, xing, id3v2, warnings, start_offset

    warnings = warn
    start_offset = offset

    f = xing = id3v2 = None
    if DEBUG:
        print("examining", file, "warnings=%i" % warnings, "offset=%i" % start_offset, "max_skip=%i" % max_skip)

    # XXX we're using StringIO because it's faster - transitionally.
    if type(file) == str:
        f = open(file, "r")
    elif hasattr(file, 'seek'):
        f = file
    else:
        print("unknown object:", file)
        return None
    f.seek(start_offset)
    search_in = f.read(max_skip)
    F = StringIO(search_in)
    F.file = file
    f.seek(start_offset)
    F.offset = start_offset

    x = syncronize(what=4)

    if not x:
        if warnings:
            print("Warning: no MP3 header found in the first", F.tell(), "bytes of file\n       : \"" + file + "\".")
        return {}

    # we're quite sure we have a valid header now
    if warnings > 1:
        print("Warning: MP3 header found at offset", f.tell() - 4)

    # fill in convenience stuff
    x = extend_frameinfo(x)

    # the xing header is normally embedded in the first valid frame
    # but sometimes it appears earlier, I don't know why. Use the 1st.
    if not xing:
        find_xing(x['framesize'])

    if xing:
        x.update(xing)
        if 'x_frames' in x:
            x['length_in_samples'] = x['x_frames'] * x['samples_per_frame']
            x['length'] = float(x['length_in_samples']) / x['sfreq']
            # should use the real file size instead of x_bytes.
            if "x_bytes" in x:
                x['bitrate'] = int((x['x_bytes'] * 8.0 / x['length']) / 100 + 0.5)
                x['bitrate'] = x['bitrate'] / 10.0
        else:
            print("strange xing=" + repr(xing))
    else:
        x['length'] = stat(file)[ST_SIZE] * 8.0 / (x['bitrate'] * 1000)
    if id3v2:
        x['id3v2'] = id3v2

    if DEBUG > 1:
        print(x)
    return x


class MP3:
    def __init__(self, file, name=None):

        self.file = None
        self.name = name
        if type(file) == str:
            self.file = open(file, "rb")
            self.name = file
        elif hasattr(file, 'seek'):
            self.file = file
        else:
            print("unknown object:", file)
            return None
